.ci/upload_function.sh

```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env bash
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+set -ex
+BINDIR=`dirname "$0"`
+PULSAR_HOME=`cd ${BINDIR}/..;pwd`
+
+NAMESPACE=default
+CLUSTER=sn-platform
+
+case ${1} in
+  java)
+    kubectl exec -n ${NAMESPACE} ${CLUSTER}-pulsar-broker-0 -- bin/pulsar-admin packages upload function://public/default/test-java-function --path /pulsar/examples/api-examples.jar --description "test java function"
+    ;;
+  py)
```

Comment:
is it possible to cover different types of packaging for python functions?
ref: https://pulsar.apache.org/docs/next/functions-package-python

Reply:
great, I will try that

Reply:
it said that `pip` is only supported in k8s runtime: https://pulsar.apache.org/docs/next/functions-package-python/#pip

Reply:
when I check python zip locally, it failed with below logs:


```
The package 'function://public/default/test-py-zip-function' downloaded to path 'exclamation.zip' successfully
shardId=0
Looking in links: exclamation/deps
Processing ./exclamation/deps/sh-1.12.14-py2.py3-none-any.whl
Installing collected packages: sh
Successfully installed sh-1.12.14
WARNING: You are using pip version 22.0.3; however, version 22.1.2 is available.
You should consider upgrading via the '/usr/bin/python3 -m pip install --upgrade pip' command.
[2022-07-20 03:12:25 +0000] [INFO] python_instance_main.py: Starting Python instance with Namespace(client_auth_params=None, client_auth_plugin=None, cluster_name='test', dependency_repository=None, expected_healthcheck_interval=-1, extra_dependency_repository=None, function_details='{"tenant":"public","namespace":"default","name":"py-function-download-zip-sample","className":"exclamation","logTopic":"persistent://public/default/py-function-logs","secretsMap":"{\\"name\\":{\\"path\\":\\"test-py-secret\\",\\"key\\":\\"username\\"},\\"pwd\\":{\\"path\\":\\"test-py-secret\\",\\"key\\":\\"password\\"}}","autoAck":true,"parallelism":1,"source":{"typeClassName":"[B","inputSpecs":{"persistent://public/default/input-download-python-zip-topic":{}},"subscriptionPosition":"EARLIEST"},"sink":{"typeClassName":"[B","topic":"persistent://public/default/output-download-python-zip-topic","forwardSourceMessageProperty":true},"resources":{"cpu":1,"ram":"1000000000"},"componentType":"FUNCTION"}', function_id='0-e549b89d-b816-446c-83bf-f36c57b91ddf', function_version='0', hostname_verification_enabled=None, install_usercode_dependencies=None, instance_id='0', logging_config_file='/pulsar/conf/functions-logging/console_logging_config.ini', logging_directory='logs/functions', logging_file='py-function-download-zip-sample-0', max_buffered_tuples='100', metrics_port=9094, port=9093, pulsar_serviceurl='pulsar://sn-platform-pulsar-broker.default.svc.cluster.local:6650', py='exclamation.zip', secrets_provider='secretsprovider.EnvironmentBasedSecretsProvider', secrets_provider_config=None, state_storage_serviceurl=None, tls_allow_insecure_connection=None, tls_trust_cert_path=None, use_tls='false')
[2022-07-20 03:12:25 +0000] [INFO] log.py: Setting up producer for log topic persistent://public/default/py-function-logs
2022-07-20 03:12:25.725 INFO  [140083473549120] ClientConnection:182 | [<none> -> pulsar://sn-platform-pulsar-broker.default.svc.cluster.local:6650] Create ClientConnection, timeout=10000
2022-07-20 03:12:25.726 INFO  [140083473549120] ConnectionPool:96 | Created connection for pulsar://sn-platform-pulsar-broker.default.svc.cluster.local:6650
2022-07-20 03:12:25.818 INFO  [140083419035392] ClientConnection:368 | [10.244.1.19:34922 -> 10.244.2.10:6650] Connected to broker
2022-07-20 03:12:25.823 INFO  [140083419035392] HandlerBase:64 | [persistent://public/default/py-function-logs, ] Getting connection from pool
2022-07-20 03:12:25.824 INFO  [140083419035392] ClientConnection:182 | [<none> -> pulsar://sn-platform-pulsar-broker-0.sn-platform-pulsar-broker.default.svc.cluster.local:6650] Create ClientConnection, timeout=10000
2022-07-20 03:12:25.824 INFO  [140083419035392] ConnectionPool:96 | Created connection for pulsar://sn-platform-pulsar-broker-0.sn-platform-pulsar-broker.default.svc.cluster.local:6650
2022-07-20 03:12:25.826 INFO  [140083419035392] ClientConnection:368 | [10.244.1.19:34924 -> 10.244.2.10:6650] Connected to broker
2022-07-20 03:12:25.830 INFO  [140083419035392] ProducerImpl:189 | [persistent://public/default/py-function-logs, ] Created producer on broker [10.244.1.19:34924 -> 10.244.2.10:6650]
2022-07-20 03:12:25.838 INFO  [140083473549120] Client:88 | Subscribing on Topic :persistent://public/default/input-download-python-zip-topic
2022-07-20 03:12:25.840 INFO  [140083473549120] HandlerBase:64 | [persistent://public/default/input-download-python-zip-topic, public/default/py-function-download-zip-sample, 0] Getting connection from pool
2022-07-20 03:12:25.844 INFO  [140083419035392] ConsumerImpl:224 | [persistent://public/default/input-download-python-zip-topic, public/default/py-function-download-zip-sample, 0] Created consumer on broker [10.244.1.19:34924 -> 10.244.2.10:6650]
[2022-07-20 03:12:25 +0000] [INFO] util.py: Failed to import class exclamation from path
[2022-07-20 03:12:25 +0000] [INFO] util.py: No module named 'sh'
Traceback (most recent call last):
  File "/pulsar/instances/python-instance/util.py", line 41, in import_class
    return import_class_from_path(from_path, full_class_name)
  File "/pulsar/instances/python-instance/util.py", line 61, in import_class_from_path
    mod = importlib.import_module(class_name)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/pulsar/exclamation/src/exclamation.py", line 1, in <module>
    import sh
ModuleNotFoundError: No module named 'sh'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/pulsar/instances/python-instance/util.py", line 46, in import_class
    return import_class_from_path(api_dir, full_class_name)
  File "/pulsar/instances/python-instance/util.py", line 61, in import_class_from_path
    mod = importlib.import_module(class_name)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/pulsar/exclamation/src/exclamation.py", line 1, in <module>
    import sh
ModuleNotFoundError: No module named 'sh'
[2022-07-20 03:12:25 +0000] [CRITICAL] python_instance.py: Could not import User Function Module exclamation
[2022-07-20 03:12:25 +0000] [ERROR] log.py: Traceback (most recent call last):
[2022-07-20 03:12:25 +0000] [ERROR] log.py:   File "/pulsar/instances/python-instance/python_instance_main.py", line 218, in <module>
[2022-07-20 03:12:25 +0000] [ERROR] log.py: main()
[2022-07-20 03:12:25 +0000] [ERROR] log.py:   File "/pulsar/instances/python-instance/python_instance_main.py", line 199, in main
[2022-07-20 03:12:25 +0000] [ERROR] log.py: pyinstance.run()
[2022-07-20 03:12:25 +0000] [ERROR] log.py:   File "/pulsar/instances/python-instance/python_instance.py", line 199, in run
[2022-07-20 03:12:25 +0000] [ERROR] log.py: raise NameError("Could not import User Function Module %s" % self.instance_config.function_details.className)
[2022-07-20 03:12:25 +0000] [ERROR] log.py: NameError
[2022-07-20 03:12:25 +0000] [ERROR] log.py: :
[2022-07-20 03:12:25 +0000] [ERROR] log.py: Could not import User Function Module exclamation
```

Reply:
looks like there is a bug in the upstream, I will fix it

Reply:
create an issue in upstream here: https://github.com/apache/pulsar/issues/16696

Reply:
both zip and pip type python has some error when specifying `--py` using a relative path, but work with an absolute path, so current test cases are using absolute path